File: Leetcode daily ( java + cpp )/cpp/2300.Successful-Pairs-of-Spells-and-Potions.cpp
// 2300. Successful Pairs of Spells and Potions

class Solution {
public:
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
        vector<int> ans;
        int n = spells.size();
        int m = potions.size();
        sort(potions.begin(),potions.end());
        for(int i=0; i<n; i++) {
            int low = 0;
            int high = m-1;
            while(low<=high) {
                int mid = (high-low)/2 + low;
                if((long long)spells[i]*potions[mid] >= success) {
                    high = mid-1;
                }
                else {
                    low = mid+1;
                }
            }
            ans.push_back(m-1-low+1);
        }
        return ans;
    }
};

File: LeetCode/2300-Successful-Pairs-of-Spells-and-Potions.java
// 2300. Successful Pairs of Spells and Potions

class Solution {
    public int lb(int[] potions, long success, int spell) {
        int n = potions.length;
        int l = 0, h = n - 1, ans = n;
        while (l <= h) {
            int m = (h - l) / 2 + l;
            if ((long) spell * potions[m] >= success) {
                ans = m;
                h = m - 1;
            }
            else {
                l = m + 1;
            }
        }
        return ans;
    }

    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        int n = spells.length, m = potions.length;
        int[] ans = new int[n];
        Arrays.sort(potions);
        for (int i = 0; i < n; i++) {
            int ind = lb(potions, success, spells[i]);
            ans[i] = m - (ind);
        }
        return ans;
    }
}
